src/Corax/Pipeline/Parsing/ScalarTransformers.cs

```diff
@@ -0,0 +1,191 @@
+                    ulong v2 = Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref bufferStart, (int)sourcePos + sizeof(ulong)));
+
+                    // We break out if non-ascii character is found
+                    if (((v1 | v2) & ParsingConstants.NonAsciiUInt64Mask) != 0)
+                        break;
+
+                    // From here on, we know we are dealing with a codepoint that is ASCII.
+
+                    // We are testing !((ch & CASE_MASK) ^ ('A' & CASE_MASK)) which means character requires transformation.
+                    if (((v1 & ParsingConstants.UppercaseUInt64Mask) ^ (0x4141414141414141 & ParsingConstants.UppercaseUInt64Mask)) != ParsingConstants.SpaceCharactersUInt64)
+                        goto LOWERCASE_ASCII;
+
+                    Unsafe.WriteUnaligned<ulong>(ref Unsafe.AddByteOffset(ref destStart, destPos), v1);
+                    sourcePos += N64;
+                    destPos += N64;
+
+                    // We are testing !((ch & CASE_MASK) ^ ('A' & CASE_MASK)) which means character requires transformation.
+                    if (((v2 & ParsingConstants.UppercaseUInt64Mask) ^ (0x4141414141414141 & ParsingConstants.UppercaseUInt64Mask)) != ParsingConstants.SpaceCharactersUInt64)
+                        goto LOWERCASE_ASCII;
+
+                    Unsafe.WriteUnaligned<ulong>(ref Unsafe.AddByteOffset(ref destStart, destPos), v2);
+                    sourcePos += N64;
+                    destPos += N64;
+
+                    continue;
+
+                    LOWERCASE_ASCII:
+                    
+                    // We perform lowercase ASCII using a loop for the ulong we detected needed the lowercase.
+                    for (int i = 0; i < 4; i++, sourcePos += 2, destPos += 2)
+                    {
+                        // dest[pos] = source[pos] | lowercaseTable[source[pos]]
+                        ref byte d1 = ref Unsafe.AddByteOffset(ref destStart, destPos);
+                        byte b1 = Unsafe.AddByteOffset(ref bufferStart, sourcePos);
+                        d1 = (byte)(b1 | Unsafe.AddByteOffset(ref lowerCaseTable, b1));
+
+                        // dest[pos+1] = source[pos+1] | lowercaseTable[source[pos+1]]
+                        ref byte d2 = ref Unsafe.AddByteOffset(ref destStart, destPos + 1);
+                        byte b2 = Unsafe.AddByteOffset(ref bufferStart, sourcePos + 1);
+                        d2 = (byte)(b2 | Unsafe.AddByteOffset(ref lowerCaseTable, b2));
+                    }
+                }
+
+                if (sourcePos == len)
+                    break;
+
+                // We know that now we must lowercase some non-ASCII character OR we don't have enough characters to work
+                // in bulk. Either way we will handle this here.
+                var opStatus = Rune.DecodeFromUtf8(source.Slice((int)sourcePos), out var rune, out int bytesConsumed);
```

Comment:
Here we decode _everything_ from `sourcePos` to the **end**. Is it correct? I have spot that after reading the above Maciej's comment which I think might be correct.

Reply:
Unless there is no guarantee that `bytesConsumed` will be everything to the end?

Reply:
It processes the current `Rune` only and will advance the pointer to the next character. It can happen that it wont end and the `opStatus` will be invalid if the data is corrupted.

Reply:
OK